File: backend/learner_api/curriculum.py
"""Read-only curriculum lookups for the training-plan builder.

Data source: normalized tables in the `curriculum` schema.

Cascade (mirrors the builder UI):
    programmes                -> curriculum.programmes
    cohorts?programme=        -> curriculum.cohorts
    groups?programme=&cohort= -> curriculum.groups
    modules?programme=        -> authored modules for the programme (NOT filtered
                                 by cohort/group — modules belong to the programme)
    weeks?module=             -> weeks for a module
    components?week=          -> components for a week

How the authoring tables link together:
    programmes(program_id, name)
        the master grid — one row per module scheduled for a group.
    modules(module_catalogue_id, programme_id, title)
        authored module content linked by programme/module ids (and, by
        title, to that group's module_name).
    weeks(id, module_catalogue_id)      -> belongs to a module
    components(id, week_id)             -> belongs to a week
Modules, weeks and components follow the module_catalogue_id / week_id chain.
"""
import json

from django.db import DatabaseError, connections
from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt


def _conn():
    return connections["enrolment"]


def _rows(sql, params=None):
    with _conn().cursor() as cur:
        cur.execute(sql, params or [])
        cols = [c[0] for c in cur.description]
        return [dict(zip(cols, row)) for row in cur.fetchall()]


def _error(message, status):
    return JsonResponse({"error": message}, status=status)


def _guard(fn):
    """Wrap a lookup so DB errors become a clean 502 instead of a 500."""
    try:
        return JsonResponse({"results": fn()})
    except DatabaseError as exc:
        return _error(f"Database error: {exc}", 502)


def programmes(request):
    if request.method != "GET":
        return _error("Method not allowed.", 405)
    return _guard(lambda: [
        r["name"]
        for r in _rows(
            "SELECT DISTINCT COALESCE(NULLIF(name, ''), NULLIF(programme_id, '')) AS name "
            "FROM curriculum.programmes "
            "WHERE COALESCE(NULLIF(name, ''), NULLIF(programme_id, '')) IS NOT NULL "
            "  AND COALESCE(is_archived, false) = false "
            "ORDER BY name"
        )
    ])


def cohorts(request):
    if request.method != "GET":
        return _error("Method not allowed.", 405)
    programme = (request.GET.get("programme") or "").strip()
    if not programme:
        return _error("programme query param is required.", 400)
    return _guard(lambda: [
        r["cohort_name"]
        for r in _rows(
            "SELECT DISTINCT cohort_name FROM curriculum.cohorts "
            "WHERE cohort_name IS NOT NULL AND cohort_name <> '' "
            "  AND (programme_name = %s OR programme_id = %s) "
            "ORDER BY cohort_name",
            [programme, programme],
        )
    ])


def groups(request):
    if request.method != "GET":
        return _error("Method not allowed.", 405)
    programme = (request.GET.get("programme") or "").strip()
    cohort = (request.GET.get("cohort") or "").strip()
    if not programme or not cohort:
        return _error("programme and cohort query params are required.", 400)
    return _guard(lambda: [
        r["group_name"]
        for r in _rows(
            "SELECT DISTINCT group_name FROM curriculum.groups "
            "WHERE group_name IS NOT NULL AND group_name <> '' "
            "  AND (programme_name = %s OR programme_id = %s) "
            "  AND (cohort_name = %s OR cohort_id = %s) "
            "ORDER BY group_name",
            [programme, programme, cohort, cohort],
        )
    ])


def modules(request):
    if request.method != "GET":
        return _error("Method not allowed.", 405)
    programme = (request.GET.get("programme") or "").strip()
    if not programme:
        return _error("programme query param is required.", 400)
    # Modules belong to the programme in the normalized curriculum schema.
    return _guard(lambda: [
        {"id": r["module_catalogue_id"], "title": r["title"] or r["module_catalogue_id"]}
        for r in _rows(
            "SELECT module_catalogue_id, title FROM curriculum.modules "
            "WHERE programme_id = %s OR programme_name = %s OR %s LIKE programme_name || ' %%' "
            "ORDER BY title",
            [programme, programme, programme],
        )
    ])


def weeks(request):
    if request.method != "GET":
        return _error("Method not allowed.", 405)
    module = (request.GET.get("module") or "").strip()
    if not module:
        return _error("module query param is required.", 400)
    return _guard(lambda: [
        {"id": r["id"], "title": r["title"] or f"Week {r['week_number']}", "weekNumber": r["week_number"]}
        for r in _rows(
            "SELECT id, week_number, title FROM curriculum.weeks "
            "WHERE module_catalogue_id = %s "
            "ORDER BY week_number, display_order",
            [module],
        )
    ])


def components(request):
    if request.method != "GET":
        return _error("Method not allowed.", 405)
    week = (request.GET.get("week") or "").strip()
    if not week:
        return _error("week query param is required.", 400)
    return _guard(lambda: [
        {
            "id": r["id"],
            "title": r["title"] or r["type"],
            "type": r["type"],
            "expectedOtjh": float(r["expected_otjh"]) if r["expected_otjh"] is not None else None,
        }
        for r in _rows(
            "SELECT id, type, title, expected_otjh FROM curriculum.components "
            "WHERE week_id = %s "
            "ORDER BY display_order",
            [week],
        )
    ])


def _ksb_profile_row(programme):
    """The KSB profile authored for a programme, or None.

    Two links exist and both are in live use, so both are tried in order of
    authority:
      1. programmes.ksb_profile_source_id — set by the curriculum structure
         wizard, holds 'profile:<KSBP-id>' (or a bare id).
      2. ksb_profiles.programme_id / .programme_name — the older direct link,
         still populated for profiles authored before the wizard existed.
    """
    programme = (programme or "").strip()
    if not programme:
        return None

    source = _rows(
        "SELECT programme_id, ksb_profile_source_id FROM curriculum.programmes "
        "WHERE name = %s OR programme_id = %s LIMIT 1",
        [programme, programme],
    )
    programme_id = (source[0].get("programme_id") or "") if source else ""
    raw_source = (source[0].get("ksb_profile_source_id") or "") if source else ""
    # Stored as 'profile:<id>' by the wizard; tolerate a bare id too.
    profile_id = raw_source.split(":", 1)[1] if ":" in raw_source else raw_source

    if profile_id:
        rows = _rows(
            "SELECT id, name, ksb_items FROM curriculum.ksb_profiles "
            "WHERE id = %s AND COALESCE(is_active, true) LIMIT 1",
            [profile_id],
        )
        if rows:
            return rows[0]

    rows = _rows(
        "SELECT id, name, ksb_items FROM curriculum.ksb_profiles "
        "WHERE COALESCE(is_active, true) "
        "  AND (programme_id = %s OR programme_name = %s) "
        "ORDER BY updated_at DESC NULLS LAST LIMIT 1",
        [programme_id or programme, programme],
    )
    return rows[0] if rows else None


_KSB_KIND = {"K": "Knowledge", "S": "Skill", "B": "Behaviour"}


def ksb_profile_for_programme(programme):
    """(standard, results) for a programme's authored KSB profile.

    Shared by the ksb_profile view and the enrolment review form, which shows the
    learner's Skills Radar answers against these same KSBs. Returns (None, []) for
    an unmapped programme so callers can say "none authored yet" rather than fail.
    """
    row = _ksb_profile_row(programme)
    if row is None:
        return None, []

    items = row.get("ksb_items")
    if isinstance(items, str):
        try:
            items = json.loads(items)
        except ValueError:
            items = []
    if not isinstance(items, list):
        items = []

    results = []
    for index, item in enumerate(items):
        if not isinstance(item, dict):
            continue
        type_ = (item.get("type") or "").strip().upper()
        number = str(item.get("code") or item.get("number") or "").strip()
        # 'K' + '1' -> 'K1'. Authored codes are sometimes already prefixed.
        code = number if number[:1].upper() in _KSB_KIND else f"{type_}{number}"
        title = (item.get("title") or item.get("description") or "").strip()
        results.append({
            "id": str(item.get("id") or f"{row['id']}-{index}"),
            "theme": (item.get("theme") or _KSB_KIND.get(type_, "")).strip(),
            "kind": _KSB_KIND.get(type_, "Knowledge"),
            "codes": [code] if code else [],
            "title": title,
            "displayOrder": item.get("displayOrder", index),
        })

    results.sort(key=lambda r: (r["displayOrder"] if isinstance(r["displayOrder"], int) else 0))
    for r in results:
        r.pop("displayOrder", None)

    return {"id": row["id"], "label": row.get("name") or programme}, results


def ksb_profile(request):
    """The KSB list a learner should self-assess against, for their programme.

        GET /learner_api/curriculum/ksb-profile/?programme=<name>

    Returns {standard: {id, label}, results: [Ksb]} shaped for the Skills Radar
    step. An unmapped programme returns an empty list rather than an error, so
    the step can say "none authored yet" instead of failing.
    """
    if request.method != "GET":
        return _error("Method not allowed.", 405)
    programme = (request.GET.get("programme") or "").strip()
    if not programme:
        return _error("programme query param is required.", 400)

    try:
        standard, results = ksb_profile_for_programme(programme)
    except DatabaseError as exc:
        return _error(f"Database error: {exc}", 502)

    return JsonResponse({"standard": standard, "results": results})


@csrf_exempt
def legacy_otjh(request):
    """Best-effort OTJH lookup for older saved training-plan components."""
    if request.method != "POST":
        return _error("Method not allowed.", 405)
    from .learner_detail import _otjh_by_legacy_title

    try:
        payload = json.loads(request.body or b"{}")
    except json.JSONDecodeError:
        return _error("Invalid JSON body.", 400)

    items = payload.get("items")
    if not isinstance(items, list):
        return _error("items must be a list of {module, week, component}.", 400)

    components_arg = [
        {"module": it.get("module"), "week": it.get("week"), "component": it.get("component")}
        for it in items if isinstance(it, dict)
    ]

    def run():
        by_key = _otjh_by_legacy_title(components_arg)
        return {
            f"{mod}|{wk}|{comp}": otjh
            for (mod, wk, comp), otjh in by_key.items()
            if otjh is not None
        }

    return _guard(run)

File: frontend/src/api/curriculum.ts
// ============================================================================
// Curriculum lookups for the training-plan builder.
// Read-only cascade sourced from normalized `curriculum` schema tables via the
// Django backend.
// ============================================================================

const BASE = '/learner_api/curriculum';

export interface CurriculumItem {
  id: string;
  title: string;
}
export interface WeekItem extends CurriculumItem {
  weekNumber: number;
}
export interface ComponentItem extends CurriculumItem {
  type: string;
  expectedOtjh: number | null;
}

async function request<T>(url: string): Promise<T> {
  let res: Response;
  try {
    res = await fetch(url, { headers: { 'Content-Type': 'application/json' } });
  } catch {
    throw new Error('Could not reach the server. Is the backend running on port 8000?');
  }
  const text = await res.text();
  const data = text ? JSON.parse(text) : null;
  if (!res.ok) throw new Error((data && data.error) || `Request failed (${res.status})`);
  return data as T;
}

const qs = (params: Record<string, string>) =>
  new URLSearchParams(params).toString();

/** Distinct programmes. */
export async function fetchProgrammes(): Promise<string[]> {
  return (await request<{ results: string[] }>(`${BASE}/programmes/`)).results;
}

/** Cohorts for a programme. */
export async function fetchCohorts(programme: string): Promise<string[]> {
  return (await request<{ results: string[] }>(`${BASE}/cohorts/?${qs({ programme })}`)).results;
}

/** Groups for a programme + cohort. */
export async function fetchGroups(programme: string, cohort: string): Promise<string[]> {
  return (await request<{ results: string[] }>(`${BASE}/groups/?${qs({ programme, cohort })}`)).results;
}

/**
 * The KSB list a learner self-assesses against, resolved from their programme's
 * authored profile in curriculum.ksb_profiles. `standard` is null when the
 * programme has no profile authored yet.
 */
export interface KsbProfileResponse {
  standard: { id: string; label: string } | null;
  results: { id: string; theme: string; kind: 'Knowledge' | 'Skill' | 'Behaviour'; codes: string[]; title: string }[];
}

export async function fetchKsbProfile(programme: string): Promise<KsbProfileResponse> {
  return request<KsbProfileResponse>(`${BASE}/ksb-profile/?${qs({ programme })}`);
}

/** Modules for a programme (independent of cohort/group). */
export async function fetchModules(programme: string): Promise<CurriculumItem[]> {
  return (await request<{ results: CurriculumItem[] }>(`${BASE}/modules/?${qs({ programme })}`)).results;
}

/** Weeks for a module. */
export async function fetchWeeks(moduleId: string): Promise<WeekItem[]> {
  return (await request<{ results: WeekItem[] }>(`${BASE}/weeks/?${qs({ module: moduleId })}`)).results;
}

/** Humanise a component type slug, e.g. 'live_session' -> 'Live session'. */
function humaniseType(type: string): string {
  if (!type) return '';
  const spaced = type.replace(/_/g, ' ');
  return spaced.charAt(0).toUpperCase() + spaced.slice(1);
}

/**
 * Components for a week.
 * Imported modules often store an unhelpful component title (e.g. "Week 1"),
 * so we surface the component TYPE and only append the title when it adds info.
 */
export async function fetchComponents(weekId: string): Promise<ComponentItem[]> {
  const results = (await request<{ results: ComponentItem[] }>(`${BASE}/components/?${qs({ week: weekId })}`)).results;
  return results.map((c) => {
    const typeLabel = humaniseType(c.type);
    const showTitle = c.title && c.title.trim().toLowerCase() !== typeLabel.toLowerCase();
    return { ...c, title: showTitle ? `${typeLabel} · ${c.title}` : typeLabel || c.title };
  });
}

export interface LegacyOtjhItem {
  module: string;
  week: string;
  component: string;
}

/**
 * expected_otjh lookup for training plans saved BEFORE the structured-plan
 * format existed — their component/week ids are client-generated, so they
 * can't be looked up by id via fetchComponents(). Resolves by (module, week,
 * component) TITLE instead, matching the same fallback the learner-facing
 * training-plan view uses. Only entries with a match are present in the
 * result map (unmatched legacy titles — e.g. a renamed/re-authored module —
 * simply have no recoverable hours).
 */
export async function fetchLegacyOtjh(items: LegacyOtjhItem[]): Promise<Record<string, number>> {
  if (items.length === 0) return {};
  let res: Response;
  try {
    res = await fetch(`${BASE}/legacy-otjh/`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ items }),
    });
  } catch {
    return {}; // best-effort — hours simply won't show for legacy items
  }
  if (!res.ok) return {};
  const data = await res.json().catch(() => null);
  return (data && data.results) || {};
}

/** Build the lookup key fetchLegacyOtjh's result map uses for one item. */
export function legacyOtjhKey(module: string, week: string, component: string): string {
  return `${module}|${week}|${component}`;
}
